core/src/main/java/fr/sncf/osrd/train/TrainPositionTracker.java

```diff
@@ -86,15 +72,53 @@ public boolean equals(Object obj) {
             return false;
 
         var other = (TrainPositionTracker) obj;
-        return trackSectionRanges.equals(other.trackSectionRanges);
+        return trackSectionPath.equals(other.trackSectionPath)
+                && headPathPosition == other.headPathPosition
+                && tailPathPosition == other.tailPathPosition;
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(trackSectionRanges);
+        return Objects.hash(trackSectionPath, headPathPosition, tailPathPosition);
     }
     // endregion
 
+    /** Resets the position tracker, but keeps the pre-computed data */
+    public TrainPositionTracker reset() {
+        headPathPosition = 0;
+        tailPathPosition = 0;
+        return this;
+    }
+
+    private static SortedDoubleMap initIntegralGrade(List<TrackSectionRange> trackSectionPath) {
+        var res = new SortedDoubleMap();
+        var sumPreviousRangeLengths = 0.;
+        var sumSlope = 0.;
+        for (var range : trackSectionPath) {
+            var grades = range.edge.forwardGradients
+                    .getValuesInRange(range.getBeginPosition(), range.getEndPosition());
+            NavigableSet<Range> keys = new TreeSet<>(grades.keySet());
+            if (range.direction == EdgeDirection.STOP_TO_START) {
+                grades = range.edge.backwardGradients
+                    .getValuesInRange(range.getBeginPosition(), range.getEndPosition());
+                keys = new TreeSet<>(grades.keySet()).descendingSet();
+            }
+
+            for (var interval : keys) {
+                var begin = sumPreviousRangeLengths + abs(range.getBeginPosition() - interval.getBeginPosition());
+                var end = sumPreviousRangeLengths + abs(range.getBeginPosition() - interval.getEndPosition());
+                var first = min(begin, end);
+                var last = max(begin, end);
+                res.put(first, sumSlope);
+                sumSlope += grades.get(interval) * interval.length();
```

Comment:
Why is the slope multiplied by the distance here ?

Reply:
Ok i just saw that you were dividing by the train length in meanTrainGrade method.
So if i understand well sumSlope is being incremented by the slope of the interval multiplied by the length of these intervals. So if first and last are in the same interval they'll have the same value.
Can we call the variable weightedSumSlope, or is it too long ?